sql: add `crdb_internal.release_series` builtin

RaduBerinde opened the pull request:
This commit adds a built-in that resolves dev versions (e.g. 23.2-10)
to the corresponding release series (e.g. 24.1). This allows us to
clean up some tests and make them stable - they will no longer need to
be adjusted when minting a release.

Informs: #112629
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113804)
<!-- Reviewable:end -->

RaduBerinde committed:
sql: add `crdb_internal.release_series` builtin

This commit adds a built-in that resolves dev versions (e.g. 23.2-10)
to the corresponding release series (e.g. 24.1). This allows us to
clean up some tests and make them stable - they will no longer need to
be adjusted when minting a release.

Informs: #112629
Release note: None

celiala reviewed (approved):
LGTM -- thank you! looks great

> Reviewed 12 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-:-Niatcqv06mHg-zyFziN:b-9aaimm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @RaduBerinde, and @rail)

___
*[`pkg/sql/sem/builtins/builtins.go` line 5013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Niast8-8e64gBE28fWU:-Niast8-8e64gBE28fWV:bhd7cwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbe481b345c5fbdc564bd73206c4e4ae2bbb2d46/pkg/sql/sem/builtins/builtins.go#L5013)):*
> ```Go
> 				}
> 				if version.Less(clusterversion.MinSupported.Version()) || clusterversion.Latest.Version().Less(version) {
> 					return nil, errors.Newf(
> ```

maybe add test case to verify that less than MinSupport or greater than Latest returns error as expected?
(e.g. maybe to builtins_test.go?)
___
*[`pkg/upgrade/upgrades/builtins_test.go` line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-NiatEoS6Hs5OJ5rq7YB:-NiatEoS6Hs5OJ5rq7YC:b7mffjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbe481b345c5fbdc564bd73206c4e4ae2bbb2d46/pkg/upgrade/upgrades/builtins_test.go#L63)):*
> ```Go
> 	sqlDB.CheckQueryResults(t, "SELECT crdb_internal.release_series(version) FROM [SHOW CLUSTER SETTING version]",
> 		[][]string{{clusterversion.Latest.ReleaseSeries().String()}})
> }
> ```

maybe add test case to verify that less than MinSupport or greater than Latest returns error as expected?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-:-NibSPA03Dhpsu3bCdaW:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @celiala, and @rail)

___
*[`pkg/sql/sem/builtins/builtins.go` line 5013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Niast8-8e64gBE28fWU:-NibSQrzCpuWnKyDZULH:byvg0q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbe481b345c5fbdc564bd73206c4e4ae2bbb2d46/pkg/sql/sem/builtins/builtins.go#L5013)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

maybe add test case to verify that less than MinSupport or greater than Latest returns error as expected?
(e.g. maybe to builtins_test.go?)
</blockquote></details>

Added to `crdb_internal`


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

andyyang890 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Late to the party but love how much cleaner this makes mixed-version logic tests!

> Reviewed 1 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-:-Nibxc3b8ifWLPf_Q5lb:brsjg7s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala and @rail)



<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala, @RaduBerinde, and @rail)

___
*[`pkg/clusterversion/dev_offset.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Nic0q0u6Jruxy3w2T25:-Nic0q0u6Jruxy3w2T26:b-oewoko) ([raw file](https://github.com/cockroachdb/cockroach/blob/15f92e56592d5498599c4bbd0136a9c9d876be1a/pkg/clusterversion/dev_offset.go#L90)):*
> ```Go
> 
> // removeDevOffset removes DevOffset from the given version, if it was applied.
> func removeDevOffset(v roachpb.Version) roachpb.Version {
> ```

nit: maybe rename this to `maybeRemoveDevOffset` for consistency


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/clusterversion/dev_offset.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Nic0q0u6Jruxy3w2T25:-NieA2is99uJ9S94Fs8r:b-jwo3y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/15f92e56592d5498599c4bbd0136a9c9d876be1a/pkg/clusterversion/dev_offset.go#L90)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: maybe rename this to `maybeRemoveDevOffset` for consistency
</blockquote></details>

`maybeRemove` suggests that it might not be removed in some cases..


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[`pkg/clusterversion/dev_offset.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Nic0q0u6Jruxy3w2T25:-NieUtMv3H_6rzPLqAuM:b5315d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/15f92e56592d5498599c4bbd0136a9c9d876be1a/pkg/clusterversion/dev_offset.go#L90)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`maybeRemove` suggests that it might not be removed in some cases..
</blockquote></details>

Hmm yeah I guess it's ambiguous, I personally interpret `maybeRemove` as "remove if it was applied, otherwise do nothing"


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113804)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[`pkg/clusterversion/dev_offset.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113804#-Nic0q0u6Jruxy3w2T25:-Niectba4_Ov2QvC4ezh:b-sgtzdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/15f92e56592d5498599c4bbd0136a9c9d876be1a/pkg/clusterversion/dev_offset.go#L90)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Maybe `withoutDevOffset`?
</blockquote></details>

Works for me! Also ok with just keeping it as-is, up to you.


<!-- Sent from Reviewable.io -->
